### ISSUE TITLE
Video lag after a long time

### ISSUE THREAD

**Issue by penguinol** [versatica/mediasoup#372]
It seems there is a bug in SeqManager:
Previous dropped packet will lead to video lag after a long time.

for example:
```#include <iostream>
#include "SeqManager.hpp"
using namespace RTC;

int main()
{
 	SeqManager<uint8_t> test;
	uint8_t i;
	uint8_t out;
	test.Input(1, i);
	for (i = 2; i < 10; ++i)
	{
		test.Drop(i);
	}
	for (i = 120; i < 140; ++i)
	{
		test.Input(i, out);
		printf("in: %d. out: %d.\r\n", i, out);
	}
}
```
output:
in: 120. out: 112.
in: 121. out: 113.
in: 122. out: 114.
in: 123. out: 115.
in: 124. out: 116.
in: 125. out: 117.
in: 126. out: 118.
in: 127. out: 119.
in: 128. out: 120.
in: 129. out: 121.
**in: 130. out: 123.
in: 131. out: 125.
in: 132. out: 127.
in: 133. out: 129.
in: 134. out: 131.
in: 135. out: 133.
in: 136. out: 135.
in: 137. out: 137.**
in: 138. out: 138.
in: 139. out: 139.

The secquence number is discontinuous between 130 ~ 137.

I think it's because we erased SeqManager::dropped set when dropped inputs older than (input-MaxValue/2), but did not minus SeqManager::base
I modified the code as follow, then it works fine:
```
	template<typename T>
	bool SeqManager<T>::Input(const T input, T& output)
	{
		auto base = this->base;

		// There are dropped inputs. Synchronize.
		if (!this->dropped.empty())
		{
			// Delete dropped inputs older than input - MaxValue/2.
			size_t droppedSize = this->dropped.size();
			auto it = this->dropped.lower_bound(input - MaxValue / 2);
			this->dropped.erase(this->dropped.begin(), it);
			this->base -= (droppedSize - this->dropped.size());
			base = this->base;

			// Check whether this input was dropped.
			it = this->dropped.find(input);
```
in: 120. out: 112.
in: 121. out: 113.
in: 122. out: 114.
in: 123. out: 115.
in: 124. out: 116.
in: 125. out: 117.
in: 126. out: 118.
in: 127. out: 119.
in: 128. out: 120.
in: 129. out: 121.
in: 130. out: 122.
in: 131. out: 123.
in: 132. out: 124.
in: 133. out: 125.
in: 134. out: 126.
in: 135. out: 127.
in: 136. out: 128.
in: 137. out: 129.
in: 138. out: 130.
in: 139. out: 131.

**Comment by ibc** [comment]
Thanks, we'll investigate this. BTW we have unit tests for this, can you please check them and say whether the should behave in a different way? or is just that those tests do not cover your scenario above?

https://github.com/versatica/mediasoup/blob/v3/worker/test/src/RTC/TestSeqManager.cpp

**Comment by jmillan** [comment]
This makes a lot of sense, nice catch @penguinol 

Yes please, it would be really nice if you could add a test that demonstrates this. It's quite clear that it fixes an existing bug.

**Comment by ibc** [comment]
I'm creating a test for this in a separate branch, please wait.

**Comment by ibc** [comment]
I've create a branch `issue-372` with a test for this. And I'm afraid it behaves correctly. This is, it behaves **exactly** as in your fixed test (but I did not change `SeqManager.cpp` code at all):

```
$ ~/src/v3-mediasoup/worker/  make test

RTC::SeqManager::Input() | [input:1]
RTC::SeqManager::Input() | result [input:1, output:1]
RTC::SeqManager::Drop() | [input:2]
RTC::SeqManager::Drop() | input dropped [input:2]
RTC::SeqManager::Drop() | [input:3]
RTC::SeqManager::Drop() | input dropped [input:3]
RTC::SeqManager::Drop() | [input:4]
RTC::SeqManager::Drop() | input dropped [input:4]
RTC::SeqManager::Drop() | [input:5]
RTC::SeqManager::Drop() | input dropped [input:5]
RTC::SeqManager::Drop() | [input:6]
RTC::SeqManager::Drop() | input dropped [input:6]
RTC::SeqManager::Drop() | [input:7]
RTC::SeqManager::Drop() | input dropped [input:7]
RTC::SeqManager::Drop() | [input:8]
RTC::SeqManager::Drop() | input dropped [input:8]
RTC::SeqManager::Drop() | [input:9]
RTC::SeqManager::Drop() | input dropped [input:9]
RTC::SeqManager::Input() | [input:120]
RTC::SeqManager::Input() | result [input:120, output:112]
RTC::SeqManager::Input() | [input:121]
RTC::SeqManager::Input() | result [input:121, output:113]
RTC::SeqManager::Input() | [input:122]
RTC::SeqManager::Input() | result [input:122, output:114]
RTC::SeqManager::Input() | [input:123]
RTC::SeqManager::Input() | result [input:123, output:115]
RTC::SeqManager::Input() | [input:124]
RTC::SeqManager::Input() | result [input:124, output:116]
RTC::SeqManager::Input() | [input:125]
RTC::SeqManager::Input() | result [input:125, output:117]
RTC::SeqManager::Input() | [input:126]
RTC::SeqManager::Input() | result [input:126, output:118]
RTC::SeqManager::Input() | [input:127]
RTC::SeqManager::Input() | result [input:127, output:119]
RTC::SeqManager::Input() | [input:128]
RTC::SeqManager::Input() | result [input:128, output:120]
RTC::SeqManager::Input() | [input:129]
RTC::SeqManager::Input() | result [input:129, output:121]
RTC::SeqManager::Input() | [input:130]
RTC::SeqManager::Input() | result [input:130, output:122]
RTC::SeqManager::Input() | [input:131]
RTC::SeqManager::Input() | result [input:131, output:123]
RTC::SeqManager::Input() | [input:132]
RTC::SeqManager::Input() | result [input:132, output:124]
RTC::SeqManager::Input() | [input:133]
RTC::SeqManager::Input() | result [input:133, output:125]
RTC::SeqManager::Input() | [input:134]
RTC::SeqManager::Input() | result [input:134, output:126]
RTC::SeqManager::Input() | [input:135]
RTC::SeqManager::Input() | result [input:135, output:127]
RTC::SeqManager::Input() | [input:136]
RTC::SeqManager::Input() | result [input:136, output:128]
RTC::SeqManager::Input() | [input:137]
RTC::SeqManager::Input() | result [input:137, output:129]
RTC::SeqManager::Input() | [input:138]
RTC::SeqManager::Input() | result [input:138, output:130]
RTC::SeqManager::Input() | [input:139]
RTC::SeqManager::Input() | result [input:139, output:131]
```

**Comment by penguinol** [comment]
It's beacuse your test case is using SeqManager<uint16_t>, but mine is using SeqManager<uint8_t>.
The input shold be larger than 32768 with uint16_t.

**Comment by penguinol** [comment]
{
			{   1,   1,   false, false },
			{   2,   0,   false, true  },
			{   3,   0,   false, true  },
			{   4,   0,   false, true  },
			{   5,   0,   false, true  },
			{   6,   0,   false, true  },
			{   7,   0,   false, true  },
			{   8,   0,   false, true  },
			{   9,   0,   false, true  },
			{   32768, 32760, false, false },
			{   32769, 32761, false, false },
			{   32770, 32762, false, false },
			{   32771, 32763, false, false },
			{   32772, 32764, false, false },
			{   32773, 32765, false, false },
			{   32774, 32766, false, false },
			{   32775, 32767, false, false },
			{   32776, 32768, false, false },
			{   32777, 32769, false, false },
			{   32778, 32770, false, false },
			{   32779, 32771, false, false },
			{   32780, 32772, false, false }
		};

**Comment by ibc** [comment]
I see. I've updated the branch so it now tests both cases. Indeed using `uint8_t` it fails. Two concerns:

1. Note that RTP seq numbers are `uint16_t` so this issue is not producing any lag. Agreed?
2. Do you know why the maths in `SeqManager.cpp` is wrong when using `uint8_t`? Where is the difference? we are using C++ templates and we are supposed to handle any input type.

**Comment by ibc** [comment]
Indeed, using your code fixes the issue with `uint8_t`, but why?

```c++
size_t droppedSize = this->dropped.size();
auto it = this->dropped.lower_bound(input - MaxValue / 2);

this->dropped.erase(this->dropped.begin(), it);
this->base -= (droppedSize - this->dropped.size());
base = this->base;
```

**Comment by penguinol** [comment]
`uint16_t` also has the problem.
Because we must let `this->dropped.erase(this->dropped.begin(), it);` really erase some items.
So input seqnum minus first dropped seqnum should larger than MaxValue/2 
This problem occurs when we stay at a low layer for about 10~20 minutes.

**Comment by ibc** [comment]
oh, I see now! In fact I've added you latest test (not yet pushed) and indeed it fails, and it works with your code. Will push it now.

**Comment by ibc** [comment]
PR #373 created!

**Comment by ibc** [comment]
merged in v3, will release a new version now.

**Comment by ibc** [comment]
Released in 3.5.6. Thanks a lot @penguinol.
